test/opt/test_fold_spec_const_op_composite.cpp

```diff
@@ -0,0 +1,922 @@
+                // Spec constants whose value can not be fully resolved should
+                // not be processed.
+                "%spec_int = OpSpecConstant %int 99",
+                // "%spec_v2int_b = OpSpecConstantComposite %v2int %signed_one %spec_int",
+              },
+              // expected
+              {
+                "%spec_v2bool = OpConstantComposite %v2bool %bool_true %bool_false",
+                "%spec_v2uint = OpConstantComposite %v2uint %unsigned_one %unsigned_one",
+                "%spec_v2int_a = OpConstantComposite %v2int %signed_one %signed_two",
+                "%spec_int = OpSpecConstant %int 99",
+                // "%spec_v2int_b = OpSpecConstantComposite %v2int %signed_one %spec_int",
+              },
+            },
+        // clang-format on
+    })));
+
+// Tests for the casting across both scalar and vector bool <-> int <-> uint.
+// int 32 <-> int 64 are not included (OpSConvert is not covered).
+INSTANTIATE_TEST_CASE_P(
+    Cast, FoldSpecConstantOpAndCompositePassTest,
+    ::testing::ValuesIn(std::vector<
+                        FoldSpecConstantOpAndCompositePassTestCase>({
+        // clang-format off
+            // int -> bool scalar
+            {
+              // original
+              {
+                "%spec_bool_t = OpSpecConstantOp %bool INotEqual %signed_three %signed_zero",
+                "%spec_bool_f = OpSpecConstantOp %bool INotEqual %signed_zero %signed_zero",
+              },
+              // expected
+              {
+                "%spec_bool_t = OpConstantTrue %bool",
+                "%spec_bool_f = OpConstantFalse %bool",
+              },
+            },
+
+            // uint -> bool scalar
+            {
+              // original
+              {
+                "%spec_bool_t = OpSpecConstantOp %bool INotEqual %unsigned_three %unsigned_zero",
+                "%spec_bool_f = OpSpecConstantOp %bool INotEqual %unsigned_zero %unsigned_zero",
+              },
+              // expected
+              {
+                "%spec_bool_t = OpConstantTrue %bool",
+                "%spec_bool_f = OpConstantFalse %bool",
+              },
+            },
+
+            // bool -> int scalar
+            {
+              // original
+              {
+                "%spec_int_one = OpSpecConstantOp %int Select %bool_true %signed_one %signed_zero",
+                "%spec_int_zero = OpSpecConstantOp %int Select %bool_false %signed_one %signed_zero",
+              },
+              // expected
+              {
+                "%spec_int_one = OpConstant %int 1",
+                "%spec_int_zero = OpConstant %int 0",
+              },
+            },
+
+            // uint -> int scalar
+            {
+              // original
+              {
+                "%spec_int_one = OpSpecConstantOp %int IAdd %unsigned_one %signed_zero",
+                "%spec_int_zero = OpSpecConstantOp %int IAdd %unsigned_zero %signed_zero",
```

Comment:
OpTypeInt says the bit pattern is always twos-complement form.  So overflow is handled correctly if you treat both as unsigned ints.  C++ specifies the overflow behaviour for unsigned ints, but not for signed ints since signed ints in C++ could be in a sign-magnitude form.